File: Codeforces/369/C.py
n,m,k2 = map(int,raw_input().split())
col = map(int,raw_input().split())

for i in xrange(n):
	col[i] -= 1

cost = []
for _ in xrange(n):
	cost.append(map(int,raw_input().split()))


dp = [[[10**12]*(m) for _ in xrange(k2+1)] for a in xrange(n)]


if col[0] == -1:
	for i in xrange(m):
		dp[0][1][i] = cost[0][i]
else:
	dp[0][1][col[0]] = 0


for i in xrange(1,n):
	for j in xrange(1,k2+1):
		if col[i] == -1:
			for k in xrange(m):
				for l in xrange(m):
					if k != l:
						dp[i][j][k] = min(dp[i][j][k],dp[i-1][j-1][l]+cost[i][k])
					else:
						dp[i][j][k] = min(dp[i][j][k],dp[i-1][j][l]+cost[i][k])
		else:
			for l in xrange(m):
				if l != col[i]:
					dp[i][j][col[i]] = min(dp[i-1][j-1][l],dp[i][j][col[i]])
				else:
					dp[i][j][col[i]] = min(dp[i-1][j][l],dp[i][j][col[i]])

ans = 10**12
for i in xrange(m):
	ans = min(ans,dp[n-1][k2][i])

if ans == 10**12:
	print -1
else:
	print ans

File: Codeforces/369/C.cpp
#include <bits/stdc++.h>
using namespace std;
#define inf 1345678912345
#define ll long long


ll dp[104][104][104];

int main(){
	int n,m,kk;
	int col[110];
	int cost[110][110];
	cin >> n >> m >> kk;
	for(int i=0;i<n;i++){
		cin >> col[i];
		col[i]--;
	}
	for(int i=0;i<n;i++){
		for(int j=0;j<m;j++){
			cin >> cost[i][j];
		}
	}
	for(int i=0;i<n;i++){
		for(int j=0;j<=n;j++){
			for(int k=0;k<m;k++){
				dp[i][j][k] = inf;
			}
		}
	}
	if (col[0] == -1){
		for(int i=0;i<m;i++){
			dp[0][1][i] = cost[0][i];
		}
	}
	else{
		dp[0][1][col[0]] = 0;
	}
	for(int i=1;i<n;i++){
		for(int j=1;j<=n;j++){
			if (col[i] == -1){
				for(int k=0;k<m;k++){
					for(int l=0;l<m;l++){
						if(l!=k){
							dp[i][j][k] = min(dp[i-1][j-1][l] + cost[i][k],dp[i][j][k]);
						}
						else{
							dp[i][j][k] = min(dp[i-1][j][l]+cost[i][k],dp[i][j][k]);
						}
					}
				}
			}
			else{
				for(int l=0;l<m;l++){
					if (l != col[i]){
						dp[i][j][col[i]] = min(dp[i-1][j-1][l],dp[i][j][col[i]]);
					}
					else{
						dp[i][j][col[i]] = min(dp[i-1][j][l],dp[i][j][col[i]]);
					}
				}
			}
		}
	}
	ll ans = inf;
	for(int i=0;i<m;i++){
		ans = min(ans,dp[n-1][kk][i]);
	}
	if(ans == inf){
		cout << -1 << endl;
	}
	else{
		cout << ans << endl;
	}
	
	return 0;
}